### ISSUE TITLE
[BlockTapestry] No support for "4x1" in switch case in method OnBeingLookedAt

### ISSUE THREAD

**Issue by GalahadLafayette** [anegostudios/VintageStory-Issues#2876]
### Game Version

v1.18.6

### Platform

Windows

### Modded

None

### SP/MP

None

### Description

During the Singleplayer game I noticed a problem with the Tapestry block, variant "Gate" which is defined as size "4x1". The problem manifests itself when the player looks at this block, we then get an error spam saying "_invalid tapestry json config - missing size attribute for size 4x1_". Additionally, lore chapter 15 (_as configured in the json file_) is not unlocked in the journal. In the case of the Multiplayer Server game, we additionally have the problem of reduced performance when anyone looks at this block (by the number of exceptions).

_A Tapestry block of size "4x1" and variant "Gate1-4" is configured by default in the json file. It is not a variant from modifications or my experiments._

I decided to diagnose this problem and checked the json file and source code for the "**BlockTapestry**" class. I noticed that the dimension "**4x1**" is not handled in the "**OnBeingLookedAt**" method in the switch. As soon as I created a Harmony path adding the missing case to the switch it was all back to normal. 

Nevertheless, I thought I would share this little discovery here.

My Harmony patch below ( I know, it's not the best solution and the code is messy, but the patch works and that's enough for me :) )

`    

           [HarmonyPatch(typeof(BlockTapestry), "OnBeingLookedAt")]
           class BlockTapestryPatch
           {
           static bool Prefix(BlockTapestry __instance, IPlayer byPlayer, BlockSelection blockSel, bool firstTick)
           {
            // Get Protected Variables and Methods
            var apiField = AccessTools.Field(typeof(BlockTapestry), "api");
            var api = (ICoreAPI)apiField.GetValue(__instance);
            var neighbours2x1Field = AccessTools.Field(typeof(BlockTapestry), "neighbours2x1");
            var neighbours2x1 = neighbours2x1Field.GetValue(__instance) as Dictionary<string, TVec2i[]>;

            var neighbours1x2Field = AccessTools.Field(typeof(BlockTapestry), "neighbours1x2");
            var neighbours1x2 = neighbours1x2Field.GetValue(__instance) as Dictionary<string, TVec2i[]>;

            var neighbours3x1Field = AccessTools.Field(typeof(BlockTapestry), "neighbours3x1");
            var neighbours3x1 = neighbours3x1Field.GetValue(__instance) as Dictionary<string, TVec2i[]>;

            var neighbours2x2Field = AccessTools.Field(typeof(BlockTapestry), "neighbours2x2");
            var neighbours2x2 = neighbours2x2Field.GetValue(__instance) as Dictionary<string, TVec2i[]>;

            var isCompleteMethod = AccessTools.Method(typeof(BlockTapestry), "isComplete");

            // Main Patch
            if (firstTick && api.Side == EnumAppSide.Server)
            {
                BlockEntityTapestry beTas = api.World.BlockAccessor.GetBlockEntity(blockSel.Position) as BlockEntityTapestry;
                if (beTas.Rotten) return true;

                string baseCode = BlockTapestry.GetBaseCode(beTas.Type);
                string size = __instance.Attributes["sizes"][baseCode].AsString();
                Dictionary<string, TVec2i[]> neighbours;

                switch (size)
                {
                    case "2x1":
                        neighbours = neighbours2x1;
                        break;
                    case "1x2":
                        neighbours = neighbours1x2;
                        break;
                    case "3x1":
                        neighbours = neighbours3x1;
                        break;
                    case "2x2":
                        neighbours = neighbours2x2;
                        break;
                    case "4x1":
                        neighbours = new Dictionary<string, TVec2i[]>()
                        {
                            { "1", new TVec2i[] { new TVec2i(1, 0, "2"), new TVec2i(2, 0, "3"), new TVec2i(3, 0, "4") } },
                            { "2", new TVec2i[] { new TVec2i(-1, 0, "1"), new TVec2i(1, 0, "3"), new TVec2i(2, 0, "4") } },
                            { "3", new TVec2i[] { new TVec2i(-2, 0, "1"), new TVec2i(-1, 0, "2"), new TVec2i(1, 0, "4") } },
                            { "4", new TVec2i[] { new TVec2i(-3, 0, "1"), new TVec2i(-2, 0, "2"), new TVec2i(-1, 0, "3") } }
                        };
                        break;
                    default:
                        throw new Exception("invalid tapestry json config - missing size attribute for size '" + size + "'");
                }

                string intComp = beTas.Type.Substring(baseCode.Length);
                TVec2i[] vecs = neighbours[intComp];

                // Use Private Method
                bool isComplete = (bool)isCompleteMethod.Invoke(__instance, new object[] { blockSel.Position, baseCode, vecs }); // From Patch
                if (isComplete)
                {

                    ModJournal jour = api.ModLoader.GetModSystem<ModJournal>();

                    if (baseCode == "schematic-c-bloody") baseCode = "schematic-c";

                    if (!jour.DidDiscoverLore(byPlayer.PlayerUID, __instance.LoreCode, __instance.GetLoreChapterId(baseCode)))
                    {
                        var splr = byPlayer as IServerPlayer;
                        jour.DiscoverLore(new LoreDiscovery() { Code = __instance.LoreCode, ChapterIds = new List<int>() { __instance.GetLoreChapterId(baseCode) } }, splr, null);
                    }
                }
            }

            return false;
        }
    }`

### How to reproduce

Simply run local serwer, look at "Tapestry Gate 4x1" and look at logs.

### Logs

```2.7.2023 01:43:57 [Server Warning] Server overloaded. A tick took 914ms to complete.
2.7.2023 01:44:02 [Server Warning] Server overloaded. A tick took 520ms to complete.
2.7.2023 01:44:05 [Server Warning] Server overloaded. A tick took 718ms to complete.
2.7.2023 01:44:06 [Server Fatal] System.Exception: invalid tapestry json config - missing size attribute for size '4x1'
  at Vintagestory.GameContent.BlockTapestry.OnBeingLookedAt (Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel, System.Boolean firstTick) [0x000d0] in <7a8165181bdf42fd92f9a2b2671b56e6>:0 
  at Vintagestory.Server.ServerSystemEntitySimulation.OnServerTick (System.Single dt) [0x00167] in <1ad860a31db44eee805a865ffbbc25e7>:0 
  at Vintagestory.Server.ServerMain.Process () [0x001d7] in <1ad860a31db44eee805a865ffbbc25e7>:0 
2.7.2023 01:44:06 [Server Fatal] System.Exception: invalid tapestry json config - missing size attribute for size '4x1'
  at Vintagestory.GameContent.BlockTapestry.OnBeingLookedAt (Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel, System.Boolean firstTick) [0x000d0] in <7a8165181bdf42fd92f9a2b2671b56e6>:0 
  at Vintagestory.Server.ServerSystemEntitySimulation.OnServerTick (System.Single dt) [0x00167] in <1ad860a31db44eee805a865ffbbc25e7>:0 
  at Vintagestory.Server.ServerMain.Process () [0x001d7] in <1ad860a31db44eee805a865ffbbc25e7>:0 
2.7.2023 01:44:06 [Server Fatal] System.Exception: invalid tapestry json config - missing size attribute for size '4x1'
  at Vintagestory.GameContent.BlockTapestry.OnBeingLookedAt (Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel, System.Boolean firstTick) [0x000d0] in <7a8165181bdf42fd92f9a2b2671b56e6>:0 
  at Vintagestory.Server.ServerSystemEntitySimulation.OnServerTick (System.Single dt) [0x00167] in <1ad860a31db44eee805a865ffbbc25e7>:0 
  at Vintagestory.Server.ServerMain.Process () [0x001d7] in <1ad860a31db44eee805a865ffbbc25e7>:0 
2.7.2023 01:44:07 [Server Fatal] System.Exception: invalid tapestry json config - missing size attribute for size '4x1'
  at Vintagestory.GameContent.BlockTapestry.OnBeingLookedAt (Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel, System.Boolean firstTick) [0x000d0] in <7a8165181bdf42fd92f9a2b2671b56e6>:0 
  at Vintagestory.Server.ServerSystemEntitySimulation.OnServerTick (System.Single dt) [0x00167] in <1ad860a31db44eee805a865ffbbc25e7>:0 
  at Vintagestory.Server.ServerMain.Process () [0x001d7] in <1ad860a31db44eee805a865ffbbc25e7>:0 
2.7.2023 01:44:07 [Server Fatal] System.Exception: invalid tapestry json config - missing size attribute for size '4x1'
  at Vintagestory.GameContent.BlockTapestry.OnBeingLookedAt (Vintagestory.API.Common.IPlayer byPlayer, Vintagestory.API.Common.BlockSelection blockSel, System.Boolean firstTick) [0x000d0] in <7a8165181bdf42fd92f9a2b2671b56e6>:0 
  at Vintagestory.Server.ServerSystemEntitySimulation.OnServerTick (System.Single dt) [0x00167] in <1ad860a31db44eee805a865ffbbc25e7>:0 
  at Vintagestory.Server.ServerMain.Process () [0x001d7] in <1ad860a31db44eee805a865ffbbc25e7>:0 
2.7.2023 01:44:09 [Server Warning] Server overloaded. A tick took 531ms to complete.
2.7.2023 01:44:11 [Server Warning] Server overloaded. A tick took 504ms to complete.
2.7.2023 01:44:18 [Server Warning] Server overloaded. A tick took 550ms to complete.
2.7.2023 01:44:20 [Server Warning] Server overloaded. A tick took 786ms to complete.
2.7.2023 01:44:24 [Server Warning] Server overloaded. A tick took 501ms to complete.
2.7.2023 01:44:25 [Server Warning] Server overloaded. A tick took 518ms to complete.
2.7.2023 01:44:27 [Server Warning] Server overloaded. A tick took 506ms to complete.
2.7.2023 01:44:29 [Server Warning] Server overloaded. A tick took 533ms to complete.```

**Comment by scubamount** [comment]
[Server Fatal] System.Exception: invalid tapestry json config - missing size attribute for size '4x1'
   at Vintagestory.GameContent.BlockTapestry.OnBeingLookedAt(IPlayer byPlayer, BlockSelection blockSel, Boolean firstTick) in C:\Users\Tyron\Documents\vintagestory\game\VSSurvivalMod\Block\BlockTapestry.cs:line 215
   at Vintagestory.Server.ServerSystemEntitySimulation.OnServerTick(Single dt) in C:\Users\Tyron\Documents\vintagestory\game\VintagestoryLib\Server\Systems\World\EntitySimulation.cs:line 165
   at Vintagestory.Server.ServerMain.Process() in C:\Users\Tyron\Documents\vintagestory\game\VintagestoryLib\Server\ServerMain.cs:line 881
that one

**Comment by Th3Dilli** [comment]
thank you, this has been fixed in #3031
